src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -1309,85 +1311,206 @@ public static unsafe int LastIndexOfAny(ref byte searchSpace, byte value0, byte
             {
-                if (LoadShort(ref first, offset) != LoadShort(ref second, offset))
+                nint offset = 0;
+                nint lengthToExamine = (nint)length - Vector<byte>.Count;
+                if (lengthToExamine > 0)
                 {
-                    goto NotEqual;
+                    do
+                    {
+                        if (LoadVector(ref first, offset) != LoadVector(ref second, offset))
+                        {
+                            goto NotEqual;
+                        }
+                        offset += Vector<byte>.Count;
+                    } while (lengthToExamine > offset);
+                }
+
+                // Do final compare as Vector<byte>.Count from end rather than start
+                Debug.Assert(lengthToExamine >= 0);
+                if (LoadVector(ref first, lengthToExamine) == LoadVector(ref second, lengthToExamine))
+                {
+                    // C# compiler inverts this test, making the outer goto the conditional jmp.
+                    goto Equal;
                 }
-                offset += sizeof(short);
-                lengthToExamine -= sizeof(short);
+
+                // This becomes a conditional jmp foward to not favor it.
+                goto NotEqual;
             }
 
-            if (lengthToExamine != IntPtr.Zero)
+#if TARGET_64BIT
+            if (Sse2.IsSupported)
             {
-                Debug.Assert((int)lengthToExamine == 1);
+                Debug.Assert(length <= sizeof(nuint) * 2);
 
-                if (Unsafe.AddByteOffset(ref first, offset) != Unsafe.AddByteOffset(ref second, offset))
+                nuint offset = length - sizeof(nuint);
+                nuint differentBits = LoadNUInt(ref first) - LoadNUInt(ref second);
+                differentBits |= LoadNUInt(ref first, offset) - LoadNUInt(ref second, offset);
+                result = (differentBits == 0);
+                goto Result;
+            }
+            else
+#endif
+            {
+                Debug.Assert(length >= sizeof(nuint));
                 {
-                    goto NotEqual;
+                    nuint offset = 0;
+                    nuint lengthToExamine = length - sizeof(nuint);
+                    if (lengthToExamine > 0)
+                    {
+                        do
+                        {
+                            // Compare unsigned so not do a sign extend mov on 64 bit
+                            if (LoadNUInt(ref first, offset) != LoadNUInt(ref second, offset))
+                            {
+                                goto NotEqual;
+                            }
+                            offset += sizeof(nuint);
+                        } while (lengthToExamine > offset);
+                    }
+
+                    // Do final compare as sizeof(nuint) from end rather than start
+                    Debug.Assert(lengthToExamine >= 0);
```

Comment:
Since _lengthToExamine_ is an unsigned data type this assertion will never trip so it can't detect errors in the logic.

Reply:
Done